### ISSUE TITLE
NNBD: Type promotion should work for `=` in the same way as `??=`

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#41172]
Example program:  

```dart
void x(int? y) {
  y = 5; // doesn't promote to int, but `??= 5` would
  y = y!; // workaround, now its promoted to int
}
```

The use case is functions like [RangeError.checkValidRange](https://api.dart.dev/stable/2.7.2/dart-core/RangeError/checkValidRange.html), which accept a nullable type, do some validation + defaulting, and return a non-nullable type. You want this function to be called and the result assigned unconditionally, so `??=` is not appropriate.

Today you are forced to do the workaround with an extra null assertion for these use cases.

cc @lrhn @leafpetersen

**Comment by jakemac53** [comment]
I am not sure if this is a language or implementation issue, please assign an area as appropriate or move the language repo :).

**Comment by scheglov** [comment]
See also https://github.com/dart-lang/sdk/issues/40088

**Comment by kevmoo** [comment]
Seems to be resolved in f724ff477b82e585a1ba571e764b01294b7f99c0

**Comment by scheglov** [comment]
61c8ac8b4661af6784b14ed9f3641e47f90bad88 implemented promotion to NonNull of the declared type in Analyzer and CFE.
